include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h

```diff
@@ -222,37 +236,51 @@ struct __parallel_transform_reduce_mid_submitter<__work_group_size, __iters_per_
           typename _ExecutionPolicy, typename _Size, typename _InitType,
           oneapi::dpl::__internal::__enable_if_device_execution_policy<_ExecutionPolicy, int> = 0, typename... _Ranges>
 auto
-__parallel_transform_reduce_mid_impl(_ExecutionPolicy&& __exec, _Size __n, const _ReduceOp& __reduce_op,
-                                     const _TransformOp& __transform_op, const _InitType& __init, _Ranges&&... __rngs)
+__parallel_transform_reduce_mid_impl(_ExecutionPolicy&& __exec, _Size __n, _ReduceOp __reduce_op,
+                                     _TransformOp __transform_op, _InitType __init, _Ranges&&... __rngs)
 {
     using _Policy = typename ::std::decay<_ExecutionPolicy>::type;
     using _CustomName = typename _Policy::kernel_name;
-    using _ReduceMainKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
-        __reduce_mid_main_kernel<::std::integral_constant<::std::uint8_t, __iters_per_work_item1>, _CustomName>>;
-    using _ReduceLeafKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
-        __reduce_mid_leaf_kernel<::std::integral_constant<::std::uint8_t, __iters_per_work_item2>, _CustomName>>;
 
-    return __parallel_transform_reduce_mid_submitter<__work_group_size, __iters_per_work_item1, __iters_per_work_item2,
-                                                     _Tp, _ReduceMainKernel, _ReduceLeafKernel>()(
-        ::std::forward<_ExecutionPolicy>(__exec), __n, __reduce_op, __transform_op, __init,
-        ::std::forward<_Ranges>(__rngs)...);
+    // The same value for __iters_per_work_item1 is currently used. Include __iters_per_work_item1 in case this changes
+    // in the future.
+    using _ReduceDeviceKernel =
+        oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<__reduce_mid_device_kernel<_CustomName>>;
+    using _ReduceWorkGroupKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
+        __reduce_mid_work_group_kernel<::std::integral_constant<::std::uint8_t, __iters_per_work_item2>, _CustomName>>;
+
+    // number of buffer elements processed within workgroup
+    constexpr _Size __size_per_work_group = __iters_per_work_item1 * __work_group_size;
+    const _Size __n_groups = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __size_per_work_group);
+    sycl::buffer<_Tp> __temp{sycl::range<1>(__n_groups)};
+
+    sycl::event __reduce_event =
+        __parallel_transform_reduce_device_kernel_submitter<__work_group_size, __iters_per_work_item1, _Tp,
+                                                            _ReduceDeviceKernel>()(
+            ::std::forward<_ExecutionPolicy>(__exec), __n, __reduce_op, __transform_op, __init, __temp,
+            ::std::forward<_Ranges>(__rngs)...);
+
+    __n = __n_groups; // Number of preliminary results from the device kernel.
+    return __parallel_transform_reduce_work_group_kernel_submitter<__work_group_size, __iters_per_work_item2, _Tp,
```

Comment:
I think ``__iters_per_work_item1`` ``__iters_per_work_item2`` deserve renaming. The name as it is today does not tell anything and it's easy to mess up with 1 and 2 numbers at the end.